### DISCUSSION TITLE
Incremental builds for projects

### DISCUSSION THREAD

**Discussion by parasyte** [vhelin/wla-dx#635] [Q&A] [answered]
What are the best options available for incremental builds in a project using wla-dx? A naive makefile quickly falls apart as a project grows and `.include` dependency depth increases. Listing all dependencies in a hand-written makefile does not scale well, putting the source of truth in two separate places (the makefile and the source code). Indiscriminately using `*.inc` as a dependency on every compilation unit is bad because changing any `inc` file will rebuild the whole project, even if it isn't used by every compilation unit.

A ton of modern build systems exist (CMake, Ninja, Bazel, SCons, and too many alternatives to mention). I don't know of any wla-dx project templates for any of these, do any exist? These tools usually rely on the compiler to provide dependency information for incremental builds, like gcc's [`-M` argument](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html#index-M).

cc65 has a [`--create-dep`](https://cc65.github.io/doc/cc65.html#option-create-dep) argument for this purpose. Is there anything in wla-dx like this? Or is the default stance to use a separate tool that recursively parses `.include` directives? Even then, it will miss out on any runtime-generated includes from macros and other non-trivial situations. But that's the best I could come up with.

**Comment by vhelin** [comment]
Well, WLA DX already supports CMake (i.e., contains CMake scripts), and there's a Visual Studio solution also. For me those are enough. And yes, there are also legacy makefiles in the "historical" folder.

I'm not personally going to make any changes/improvements to these as they work and I know very little about build scripts. But if someone wants to make them better, feel free to create a pull request. :)

**Reply by parasyte** [reply]
I'm not talking about building wla-dx itself; I'm talking about using wla-dx to compile a 6502 assembly project, orchestrated with some build system. @cr1901's answer might point me in the right direction.

**Comment by cr1901** [comment, marked as answer]
@vhelin I think OP is asking "does WLA-DX support the `-M` flag that Unix C compilers do to generate dependency information at build time"? And, according to the help, it does:

```
$ wla-6502 -h
----------------------------------------------------------------------
---                WLA 6502 Macro Assembler v10.6a                 ---
----------------------------------------------------------------------
                Programmed by Ville Helin in 1998-2008
        In GitHub since 2014: https://github.com/vhelin/wla-dx
                   $VER: wla-6502 10.6a (5.8.2023)

USAGE: C:\msys64\mingw64\bin\wla-6502.exe [OPTIONS] <OUTPUT> <ASM FILE>

OPTIONS:
-c  Continue parsing after an error
-d  Disable WLA's ability to calculate A-B where A and B are labels
-h  Assume all label references are 16-bit by default (size hints
    still work)
-i  Add list file information
-k  Keep empty sections
-M  Output makefile rules
-MP Create a phony target for each dependency other than the main file,
    use this with -M
-MF <FILE> Specify a file to write the dependencies to, use with -M
-q  Quiet
-s  Don't create _sizeof_* and _paddingof_* definitions
-t  Test assemble
-v  Verbose messages
-v1 Verbose messages (only discard sections)
-v2 Verbose messages (-v1 plus short summary)
-w  Require labels to end in a colon
-x  Extra compile time labels and definitions
-I <DIR>  Include directory
-D <DEF>  Declare definition

OUTPUT:
-o <FILE>  Output object file
-l <FILE>  Output library file

EXAMPLES:
C:\msys64\mingw64\bin\wla-6502.exe -M -t -o main.o main.asm
C:\msys64\mingw64\bin\wla-6502.exe -D VER=1 -D TWO=2 -v -o main.o main.asm
```

Tho I've never used `-M` flags all that successfully in my C projects LOL...

**Reply by vhelin** [reply]
Shows how little I understand build system related things. :D

**Reply by parasyte** [reply]
Awesome, thank you! Maybe I should have just checked the help text. Is it just me, or is the [documentation](https://wla-dx.readthedocs.io/en/latest/) missing all of the cli argument information?

**Reply by Neui** [reply]
At the bottom there are the man pages which reference `-M` but it appears it hasn't been updated with the other options.
